src/rgw/rgw_sal_motr.h

```diff
@@ -40,6 +43,177 @@ class MotrStore;
 #define RGW_MOTR_BUCKET_HD_IDX_NAME   "motr.rgw.bucket.headers"
 //#define RGW_MOTR_BUCKET_ACL_IDX_NAME  "motr.rgw.bucket.acls"
 
+// Implement watch-notify using Motr FDMI.
+// (1) A set of global indices, notify_watch_indices
+// (2) For each object to be watched, hash(obj_name) is used to pick which
+//     index in notifiy_watch_indices to insert a new key/value record,
+//     key = unique_fid(obj_name), value = marker (for fdmi filter) + notification msg.
+//     If multiple rgw instances are trying to write the same objects at the same time,
+//     we assume that Motr index is updated atomically.
+// (3) When the index is written, an FDMI record with a special marker is generated and
+//     picked up by the filter and delivered to our fdmi application, watcher.
+// (4) The embedded notification message is then processed to get object name and index
+//     opcode. These info are sent to cache layer.
+// (5) Update cache item accordingly and release notification(fdmi record).
+
+#define RGW_MOTR_CACHE_FDMI_FILTER_MARKER "rgw.motr.cache.fdmi.marker"
+
+// Notification message.
+class MotrWatchNotifyMsg {
+protected:
+  // The marker is used by FDMI to filter out the notification.
+  std::string marker;
+  std::string sender;
+
+public:
+  MotrWatchNotifyMsg(std::string _marker, std::string _sender) : marker(_marker), sender(_sender) {}
+  MotrWatchNotifyMsg(std::string _marker) : marker(_marker) {}
+  MotrWatchNotifyMsg() {}
+  virtual ~MotrWatchNotifyMsg() = default;
+  virtual void encode(bufferlist& bl) const {
+    // Add raw marker string at the begining to let FDMI filter
+    // catch it easily and make the header readable. Encoding marker
+    // with sender like the uncommented code below also works (and has
+    // been tested).
+    uint32_t marker_len = marker.length();
+    char len_str[16];
+    snprintf(len_str, sizeof(len_str), "%08d", marker_len);
+    std::string head = std::string(len_str) + '.' + marker;
+    bl.append(head);
+
+    ENCODE_START(2, 2, bl);
+    //encode(marker, bl);
+    encode(sender, bl);
+    ENCODE_FINISH(bl);
+  }
+  virtual void decode(bufferlist::const_iterator& bl) {
+    uint32_t marker_len;
+    char len_str[16];
+    bl.copy(8, (char *)len_str);
+    marker_len = std::stoul(len_str, nullptr, 10);
+    marker.clear();
+    bl.copy(marker_len + 1, marker);
+    marker.erase(0, 1);
+
+    DECODE_START(2, bl);
+    //decode(marker, bl);
+    decode(sender, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(MotrWatchNotifyMsg)
+
+class MotrCacheNotif : public MotrWatchNotifyMsg {
```

Comment:
What's the purpose of this class? Please add some doc like you did for `MotrWatcher` below.

Reply:
This is a good description. Can it be added as a comment?